File: built-in plugins dsp/mb-vocal-harmony.cpp
/**
 * MB Harmony Engine
 * Category : effect
 * Type     : vocal
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Intelligent harmony generation
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_VOCAL_HARMONY_H
#define MB_VOCAL_HARMONY_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbVocalHarmony : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-vocal-harmony";
    static constexpr const char* PLUGIN_NAME    = "MB Harmony Engine";
    static constexpr const char* PLUGIN_TYPE    = "vocal";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float voices = 2f;  // range [1, 4]
    float interval1 = 3f;  // range [-12, 12]
    float interval2 = 5f;  // range [-12, 12]
    float mix = 0.5f;  // range [0, 1]
    };

    MbVocalHarmony() = default;
    ~MbVocalHarmony() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.voices = std::clamp(params.voices, 1f, 4f);
        params.interval1 = std::clamp(params.interval1, -12f, 12f);
        params.interval2 = std::clamp(params.interval2, -12f, 12f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Harmony Engine
        return input;
    }
};

#endif // MB_VOCAL_HARMONY_H

File: built-in plugins dsp/mb-vocal-harmony.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbVocalHarmonyPlugin: PluginDefinition = {
  id: "mb-vocal-harmony",
  slug: "mb-vocal-harmony",
  name: "MB Harmony Engine",
  category: "effect",
  type: "vocal",
  version: "1.0.0",
  description: "Intelligent harmony generation",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "voices",
      name: "Voices",
      type: "float",
      defaultValue: 2,
      minValue: 1,
      maxValue: 4,
      automatable: false,
    },
    {
      id: "interval1",
      name: "Interval 1",
      type: "float",
      defaultValue: 3,
      minValue: -12,
      maxValue: 12,
      automatable: true,
    },
    {
      id: "interval2",
      name: "Interval 2",
      type: "float",
      defaultValue: 5,
      minValue: -12,
      maxValue: 12,
      automatable: true,
    },
    {
      id: "mix",
      name: "Mix",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { voices: 2, interval1: 3, interval2: 5, mix: 0.5 },
};

export default MbVocalHarmonyPlugin;
